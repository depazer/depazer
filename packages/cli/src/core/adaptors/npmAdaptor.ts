import type { ModuleGraph, ModuleObject } from '@/types/moduleGraph'

/**
 * @desc npm module graph generator
 * @param root project root
 */
export async function npmAdaptor(
  _root: string,
  _moduleObject: ModuleObject,
  _depth: number
  // @ts-ignore
): Promise<ModuleGraph> {}
