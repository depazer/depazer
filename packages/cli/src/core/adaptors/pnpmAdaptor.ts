import type { ModuleGraph, ModuleObject } from '@/types/moduleGraph'

export async function pnpmAdaptor(
  _root: string,
  _moduleObject: ModuleObject,
  _depth: number
  // @ts-ignore
): Promise<ModuleGraph> {}
