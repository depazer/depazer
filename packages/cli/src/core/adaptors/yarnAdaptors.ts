import type { ModuleGraph, ModuleObject } from '@/types/moduleGraph'

export async function yarnAdaptor(
  _root: string,
  _moduleObject: ModuleObject,
  _depth: number
  // @ts-ignore
): Promise<ModuleGraph> {}
