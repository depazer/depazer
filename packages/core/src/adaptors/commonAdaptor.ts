import { hasFile, readPackageJSON } from '@depazer/shared'
import { resolve } from 'node:path'

import type { DependencyTree, SupportedPackageManager } from '@/types/dependencyDigraph'

const defaultDependency = {
  version: 'unknown',
  dependencies: []
}

const SHARED = {
  markSet: new Set<string>(),
  maxDepth: Infinity,
  packageManager: 'npm',
  pnpmRoot: ''
}

/**
 * @desc 通用模块解析器
 * @test pnpm - 8.6.12 npm - 9.6.7 yarn - 1.22.19 bun - 0.7.3
 * @param packageManager
 * @param root
 * @param moduleObject
 * @param depth
 * @returns
 */
export async function commonAdaptor(
  packageManager: SupportedPackageManager,
  root: string,
  moduleObject: DependencyTree,
  depth: number
): Promise<DependencyTree> {
  if (depth <= 1) return moduleObject

  SHARED.markSet.clear()
  SHARED.maxDepth = depth
  SHARED.packageManager = packageManager
  if (packageManager === 'pnpm') SHARED.pnpmRoot = resolve(root, 'node_modules/.pnpm')

  moduleObject.dependencies = await Promise.all(
    moduleObject.dependencies.map(async ({ name }) => {
      return await resolveModule(root, root, name, 2)
    })
  )

  moduleObject.devDependencies = await Promise.all(
    moduleObject.devDependencies!.map(async ({ name }) => {
      return await resolveModule(root, root, name, 2)
    })
  )
  return moduleObject
}

/**
 * @desc 搜索模块的package.json文件
 * @param root 项目根目录 查找路径上限
 * @param currentPath 当前模块路径
 * @param dependency 模块名
 * @param depth 当前模块深度
 * @returns 模块信息
 */
async function resolveModule(
  root: string,
  currentPath: string,
  dependency: string,
  depth: number
): Promise<DependencyTree> {
  const dependencyPath = resolve(currentPath, 'node_modules', dependency)
  const packageJSONPath = resolve(dependencyPath, 'package.json')

  if ((await hasFile(packageJSONPath)) === false) {
    /** @desc 防止文件夹越界，超出根目录 */
    if (currentPath === root) {
      const isPNPM = SHARED.packageManager === 'pnpm'

      return isPNPM && (await hasFile(SHARED.pnpmRoot, `node_modules/${dependency}/package.json`))
        ? resolveModule(root, SHARED.pnpmRoot, dependency, depth)
        : { name: dependency, ...defaultDependency }
    }
    /** @desc 查找父级目录 */
    return resolveModule(root, resolve(currentPath, '..'), dependency, depth)
  }

  const { name, version, dependencies } = await readPackageJSON(packageJSONPath)

  if (SHARED.markSet.has(packageJSONPath)) {
    return { ...defaultDependency, name: dependency, version }
  } else {
    SHARED.markSet.add(packageJSONPath)
  }

  return {
    ...defaultDependency,
    name,
    version,
    dependencies: await generateModuleObject(root, dependencyPath, dependencies, depth)
  }
}

/**
 * @desc 生成模块依赖对象, 若超出最大深度则返回package.json依赖的书面版本, 不再递归
 * @param root 同上
 * @param currentPath 同上
 * @param dependencies 模块依赖
 * @param depth 同上
 * @returns 模块依赖对象
 */
async function generateModuleObject(
  root: string,
  currentPath: string,
  dependencies: Record<string, string>,
  depth: number
) {
  return depth >= SHARED.maxDepth
    ? Object.entries(dependencies).map(([name, version]) => ({
        ...defaultDependency,
        name,
        version
      }))
    : await Promise.all(
        Object.keys(dependencies).map(async (name) => {
          return await resolveModule(root, currentPath, name, depth + 1)
        })
      )
}
