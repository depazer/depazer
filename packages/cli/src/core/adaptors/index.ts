import { npmAdaptor } from './npmAdaptor'
import { pnpmAdaptor } from './pnpmAdaptor'

export default {
  npm: npmAdaptor,
  pnpm: pnpmAdaptor,
  yarn: npmAdaptor
}
