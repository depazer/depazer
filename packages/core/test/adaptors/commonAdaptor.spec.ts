import { commonAdaptor } from '@/adaptors/commonAdaptor'
import { describe, test, vi } from 'vitest'

describe('commonAdaptor', () => {
  const { mockedHasFile } = vi.hoisted(() => ({
    mockedHasFile: vi.fn()
  }))

  vi.mock('@depazer/shared', async (getSourceMod) => {
    const sourceMod = await getSourceMod<Record<string, Function>>()
    return {
      ...sourceMod,
      hasFile: mockedHasFile,
      readPackageJSON: async () => ({
        name: 'test',
        version: '1.0.0',
        dependencies: { 'test-dep': '1.0.0' },
        devDependencies: { 'test-dev-dep': '1.0.0' }
      })
    }
  })

  test('npm & depth 1', async ({ expect }) => {
    const res = await commonAdaptor(
      'npm',
      '.',
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [{ name: 'test-dep', version: '1.0.0', dependencies: [] }],
        devDependencies: [{ name: 'test-dev-dep', version: '1.0.0', dependencies: [] }]
      },
      1
    )

    expect(res).toMatchSnapshot()
  })

  test('npm & depth 5', async ({ expect }) => {
    vi.mocked(mockedHasFile).mockResolvedValue(true)

    const res = await commonAdaptor(
      'npm',
      '.',
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [
          { name: 'test-dep', version: '1.0.0', dependencies: [] },
          { name: 'test', version: '1.0.0', dependencies: [] }
        ],
        devDependencies: [
          { name: 'test-dep', version: '1.0.0', dependencies: [] },
          { name: 'test-dev-dep', version: '1.0.0', dependencies: [] }
        ]
      },
      5
    )

    expect(res).toMatchSnapshot()
  })

  test('yarn & depth 1', async ({ expect }) => {
    const res = await commonAdaptor(
      'yarn',
      '.',
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [{ name: 'test-dep', version: '1.0.0', dependencies: [] }],
        devDependencies: [{ name: 'test-dev-dep', version: '1.0.0', dependencies: [] }]
      },
      1
    )

    expect(res).toMatchSnapshot()
  })

  test('pnpm & depth 1', async ({ expect }) => {
    const res = await commonAdaptor(
      'pnpm',
      '.',
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [{ name: 'test-dep', version: '1.0.0', dependencies: [] }],
        devDependencies: [{ name: 'test-dev-dep', version: '1.0.0', dependencies: [] }]
      },
      1
    )

    expect(res).toMatchSnapshot()
  })

  test('pnpm & depth 4 & search .pnpm/node_modules', async ({ expect }) => {
    vi.mocked(mockedHasFile).mockResolvedValueOnce(true)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValue(true)

    const res = await commonAdaptor(
      'pnpm',
      process.cwd(),
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [{ name: 'test-dep', version: '1.0.0', dependencies: [] }],
        devDependencies: []
      },
      3
    )

    expect(res).toMatchSnapshot()
  })

  test('pnpm & depth 4 & have multiple in root/node_modules & search .pnpm/node_modules', async ({
    expect
  }) => {
    vi.mocked(mockedHasFile).mockResolvedValueOnce(true)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(true)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValue(true)

    const res = await commonAdaptor(
      'pnpm',
      process.cwd(),
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [{ name: 'test-dep', version: '1.0.0', dependencies: [] }],
        devDependencies: []
      },
      3
    )

    expect(res).toMatchSnapshot()
  })

  test('pnpm & depth 5 & with same dependency', async ({ expect }) => {
    vi.mocked(mockedHasFile).mockResolvedValueOnce(true)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValueOnce(false)
    vi.mocked(mockedHasFile).mockResolvedValue(true)

    const res = await commonAdaptor(
      'pnpm',
      '.',
      {
        name: 'test',
        version: '1.0.0',
        dependencies: [
          { name: 'test-dep', version: '1.0.0', dependencies: [] },
          { name: 'test', version: '1.0.0', dependencies: [] }
        ],
        devDependencies: [
          { name: 'test-dep', version: '1.0.0', dependencies: [] },
          { name: 'test-dev-dep', version: '1.0.0', dependencies: [] }
        ]
      },
      5
    )

    expect(res).toMatchSnapshot()
  })
})
