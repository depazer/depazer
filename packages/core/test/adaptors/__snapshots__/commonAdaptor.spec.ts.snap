// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`commonAdaptor > npm & depth 1 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [],
      "name": "test-dev-dep",
      "version": "1.0.0",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > npm & depth 5 1`] = `
{
  "dependencies": [
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "1.0.0",
    },
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > pnpm & depth 1 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [],
      "name": "test-dev-dep",
      "version": "1.0.0",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > pnpm & depth 4 & have multiple in root/node_modules & search .pnpm/node_modules 1`] = `
{
  "dependencies": [
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [],
              "name": "test-dep",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > pnpm & depth 4 & search .pnpm/node_modules 1`] = `
{
  "dependencies": [
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [],
              "name": "test-dep",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > pnpm & depth 5 & with same dependency 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "unknown",
    },
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
    {
      "dependencies": [
        {
          "dependencies": [
            {
              "dependencies": [
                {
                  "dependencies": [
                    {
                      "dependencies": [],
                      "name": "test-dep",
                      "version": "1.0.0",
                    },
                  ],
                  "name": "test",
                  "version": "1.0.0",
                },
              ],
              "name": "test",
              "version": "1.0.0",
            },
          ],
          "name": "test",
          "version": "1.0.0",
        },
      ],
      "name": "test",
      "version": "1.0.0",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;

exports[`commonAdaptor > yarn & depth 1 1`] = `
{
  "dependencies": [
    {
      "dependencies": [],
      "name": "test-dep",
      "version": "1.0.0",
    },
  ],
  "devDependencies": [
    {
      "dependencies": [],
      "name": "test-dev-dep",
      "version": "1.0.0",
    },
  ],
  "name": "test",
  "version": "1.0.0",
}
`;
